threedi_api_client/files.py

```diff
@@ -0,0 +1,129 @@
+import logging
+import os
+import re
+from typing import BinaryIO
+from typing import Tuple
+from urllib.parse import urlparse
+
+import urllib3
+
+from openapi_client.exceptions import ApiException
+
+CONTENT_RANGE_REGEXP = re.compile(r"^bytes (\d+)-(\d+)/(\d+|\*)$")
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_pool(retries: int = 3, backoff_factor: int = 1) -> urllib3.PoolManager:
+    """Create a PoolManager with a retry policy.
+
+    The default retry policy has 3 retries with 1, 2, 4 second intervals.
+
+    Args:
+        retries: Total number of retries per request
+        backoff_factor: Multiplier for retry delay times (1, 2, 4, ...)
+    """
+    return urllib3.PoolManager(
+        retries=urllib3.util.retry.Retry(retries, backoff_factor=backoff_factor)
+    )
+
+
+def download_file(
+    url: str,
+    target: str,
+    chunk_size: int = 16777216,
+    timeout: float = 5.0,
+    pool: urllib3.PoolManager = None,
+) -> Tuple[str, int]:
+    """Download a file to a specified path on disk.
+
+    It is assumed that the file server supports multipart downloads.
+
+    Args:
+        url: the url to retrieve
+        target: the location to copy to. if this is a directory, a
+            filename is generated from the filename in the url.
+        chunk_size: numer of bytes per request. Default: 16MB.
+        timeout: the total timeout in seconds
+        pool: an optional PoolManager. if not given, a default one will be
+            created with a retry policy of 3 retries after 1, 2, 4 seconds.
+
+    Returns:
+        tuple of file path, total file size in bytes
+
+    If `path` already exists, it's overwritten. This will fail if the remote
+    server does not support range requests. Retries failed part requests 3
+    times after 1, 2 and 4 seconds.
+    """
+    # if it is a directory, take the filename from the url
+    if os.path.isdir(target):
+        filename = urlparse(url)[2].rsplit("/", 1)[-1]
+        target = os.path.join(target, filename)
+
+    # open the file
+    with open(target, "wb") as fileobj:
+        size = download_fileobj(url, fileobj, chunk_size, timeout, pool)
```

Comment:
Would be nice to have some user friendly error message in case anything goes wrong (I'm thinking IOError/OSError or whatever `download_fileobj` could raise).